api/http/offlinegate/offlinegate_wrapper.go

```diff
@@ -0,0 +1,58 @@
+package offlinegate
+
+import (
+	"log"
+	"net/http"
+	"strings"
+	"sync"
+
+	httperror "github.com/portainer/libhttp/error"
+)
+
+const RedirectReasonAdminInitTimeout string = "AdminInitTimeout"
+
+// OfflineGateWrapper is an entity used to maintain the administrator initialization status
+type OfflineGateWrapper struct {
+	lock              *sync.Mutex
+	adminInitDisabled bool
+}
+
+// NewOfflineGateWrapper creates a new gate wrapper
+func NewOfflineGateWrapper(timeoutCh chan interface{}) *OfflineGateWrapper {
+	gateWrapper := &OfflineGateWrapper{
+		lock:              &sync.Mutex{},
+		adminInitDisabled: false,
+	}
+
+	go func() {
+		<-timeoutCh
+		log.Println("[INFO] Please restart Portainer instance and initialize the administrator")
+		gateWrapper.DisableInstance()
+	}()
+	return gateWrapper
+}
+
+func (o *OfflineGateWrapper) DisableInstance() {
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	o.adminInitDisabled = true
+}
+
+func (o *OfflineGateWrapper) WasDisabled() bool {
```

Comment:
This function needs locking to avoid a data race, I suggest using an RWMutex so it doesn't become a bottleneck.

You can verify it by running `go test -race .` for this package.

Reply:
Good advice. I will check it and update.

Reply:
fixed